Mutex.php

```diff
@@ -0,0 +1,213 @@
+     * This method will initialize the [[redis]] property to make sure it refers to a valid redis connection.
+     *
+     * @throws InvalidConfigException if [[redis]] is invalid.
+     */
+    public function init()
+    {
+        parent::init();
+        $this->redis = Instance::ensure($this->redis, Connection::className());
+
+        if ($this->keyPrefix === null) {
+            $this->keyPrefix = substr(md5(Yii::$app->id), 0, 5);
+        }
+    }
+
+    /** @inheritdoc */
+    protected function acquireLock($name, $timeout = 0)
+    {
+        list($lockKey, $lockValue) = $this->addLock($name);
+
+        if (null === $lockKey) {
+            return false;
+        }
+
+        /**
+         * Set lock command
+         *
+         * @return array|bool|null|string
+         */
+        $setLock = function () use ($name, $lockKey, $lockValue) {
+            return $this->redis->executeCommand('SET', [$lockKey, $lockValue, 'PX', $this->lockExpire, 'NX']);
+        };
+
+        if ($setLock()) {
+            return true;
+        }
+
+        while ($timeout > 0) {
+            usleep($this->lockWaitSleep * 1000);
+            $timeout -= $this->lockWaitSleep;
+
+            if ($setLock()) {
+                return true;
+            }
+        }
+
+        $this->deleteLock($name);
+
+        return false;
+    }
+
+    /** @inheritdoc */
+    protected function releaseLock($name)
+    {
+        list($lockKey, $lockValue) = $this->getLock($name);
+
+        if (null === $lockKey) {
+            return false;
+        }
+
+        $luaScript = 'if redis.call("GET", KEYS[1]) == ARGV[1] then
+                        return redis.call("DEL", KEYS[1])
+                    else
+                        return 0
+                    end';
+
+        if ($this->redis->executeCommand('EVAL', [$luaScript, 1, $lockKey, $lockValue])) {
+            $this->deleteLock($name);
+
+            return true;
+        }
+
+        return false;
+    }
+
+    /**
+     * Generates redis key for a lock name.
+     *
+     * @param string $lockName
+     *
+     * @return string
+     */
+    protected function getLockKey($lockName)
+    {
+        return $this->keyPrefix . $lockName;
+    }
+
+    /**
+     * Gets lock from the local map
+     *
+     * @param string $name
+     *
+     * @return array|null lock's [keyName, keyValue] pair for redis or null if there is no such lock
+     */
+    protected function getLock($name)
+    {
+        return isset($this->lockMap[$name]) ? [$this->getLockKey($name), $this->lockMap[$name]] : null;
+    }
+
+    /**
+     * Adds lock to the local map
+     *
+     * @param string $name
+     *
+     * @return array|null lock's [keyName, keyValue] pair for redis or null if it's already exists in $lockMap
+     */
+    protected function addLock($name)
+    {
+        if (isset($this->lockMap[$name])) {
+            return null;
+        }
+
+        $lockValue            = Yii::$app->security->generateRandomString();
```

Comment:
Fomatting should be fixed to match what we have in Yii: https://github.com/yiisoft/yii2-coding-standards